### PR TITLE
Fix chunk buffer

### PR DESCRIPTION
**Pull request by pvizeli**
Replace #253

### CODE REVIEW

File: snitun/server/run.py [hunk 1]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            return
         _LOGGER.debug("No responsible worker for %s", hostname)
 
-        self._close_socket(con)
-        return None
+        client.close_socket()
+        return
 
     @staticmethod
     def _close_socket(con: socket.socket, shutdown: bool = True) -> None:
```
Comment (ludeeus) [inline snitun/server/run.py:365]: This is now unused

File: snitun/server/run.py [hunk 2]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+    ) -> None:
         """Process connection & helo."""
         try:
-            data += con.recv(MAX_READ_SIZE)
+            client.buffer += client.sock.recv(MAX_READ_SIZE)
         except OSError as err:
             _LOGGER.warning("Receive fails: %s", err)
-            self._close_socket(con, shutdown=False)
-            return None
+            client.close_socket(shutdown=False)
```
Comment (ludeeus) [inline snitun/server/run.py:318]: We need to unregister before closing here.

File: snitun/server/run.py [hunk 3]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            return
 
         # No data received
-        if not data:
-            self._close_socket(con)
-            return None
+        if not client.buffer:
+            client.close_socket()
```
Comment (ludeeus) [inline snitun/server/run.py:323]: We need to unregister before closing here.

File: snitun/server/run.py [hunk 4]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            return
 
         # Peer connection
-        if data.startswith(b"gA"):
-            next(workers_lb).handover_connection(con, data)
-            _LOGGER.debug("Handover new peer connection: %s", data)
-            return None
+        if client.buffer.startswith(b"gA"):
```
Comment (ludeeus) [inline snitun/server/run.py:327]: We need to unregister before handing over here.

File: snitun/server/run.py [hunk 5]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            return
 
         # TLS/SSL connection
-        if data[0] != 0x16:
-            _LOGGER.warning("No valid ClientHello found: %s", data)
-            self._close_socket(con)
-            return None
+        if client.buffer[0] != 0x16:
+            _LOGGER.warning("No valid ClientHello found: %s", client.buffer)
+            client.close_socket()
```
Comment (ludeeus) [inline snitun/server/run.py:336]: We need to unregister before closing here.

File: snitun/server/run.py [hunk 6]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+        # Get Hostname
         try:
-            hostname = parse_tls_sni(data)
+            hostname = parse_tls_sni(client.buffer)
         except ParseSNIIncompleteError:
-            return data
+            # Check Buffer Size
+            if len(client.buffer) >= MAX_BUFFER_SIZE:
+                _LOGGER.warning("Connection %d exceed buffer size", client.fileno)
+                client.close_socket()
```
Comment (ludeeus) [inline snitun/server/run.py:346]: We need to unregister before closing here.

File: snitun/server/run.py [hunk 7]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            return
         except ParseSNIError:
             _LOGGER.warning("Receive invalid ClientHello on public Interface")
-            return None
+            client.close_socket()
```
Comment (ludeeus) [inline snitun/server/run.py:350]: We need to unregister before closing here.

File: snitun/server/run.py [hunk 8]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            return
+
+        # Distribute to child worker
         for worker in self._workers:
             if not worker.is_responsible_peer(hostname):
                 continue
-            worker.handover_connection(con, data, sni=hostname)
+            worker.handover_connection(client.sock, client.buffer, sni=hostname)
```
Comment (ludeeus) [inline snitun/server/run.py:357]: We need to unregister before handing it over here.

File: snitun/server/run.py [hunk 9]
```diff
@@ -303,53 +304,62 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            client.close = True
 
             _LOGGER.info("Handover %s to %s", hostname, worker.name)
-            return None
+            return
         _LOGGER.debug("No responsible worker for %s", hostname)
 
-        self._close_socket(con)
-        return None
+        client.close_socket()
```
Comment (ludeeus) [inline snitun/server/run.py:364]: We need to unregister before closing here.

File: snitun/server/run.py [hunk 10]
```diff
@@ -18,7 +18,7 @@
 import attr
 
 from ..exceptions import ParseSNIIncompleteError
-from ..utils.server import MAX_READ_SIZE
+from ..utils.server import MAX_READ_SIZE, MAX_BUFFER_SIZE
```
Comment (ludeeus) [inline snitun/server/run.py:21]:
```suggestion
from ..utils.server import MAX_BUFFER_SIZE, MAX_READ_SIZE
```

File: snitun/server/run.py [hunk 11]
```diff
@@ -231,20 +252,8 @@ def stop(self) -> None:
...
-            """Register partial data."""
-            if current := partial.get(fileno):
-                current.data = data
-                current.count += 1
-            else:
-                partial[fileno] = current = PartialData(data=data)
-
-            return len(current.data) < MAX_READ_SIZE and current.count < 4
 
         _LOGGER.warning("Server started, fd: %s", fd_server)
```
Comment (ludeeus) [inline snitun/server/run.py:258]:
```suggestion
```

File: snitun/server/run.py [hunk 12]
```diff
@@ -303,58 +308,59 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
+            client.close_socket()
+            return
+
+        # Distribute to child worker
         for worker in self._workers:
             if not worker.is_responsible_peer(hostname):
                 continue
-            worker.handover_connection(con, data, sni=hostname)
+            worker.handover_connection(client.sock, client.buffer, sni=hostname)
+            client.soft_close()
```
Comment (ludeeus) [inline snitun/server/run.py:359]: Shouldn't this happen before the handover?
Like before?

File: snitun/server/run.py [hunk 13]
```diff
@@ -303,58 +308,59 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
 
         # Peer connection
-        if data.startswith(b"gA"):
-            next(workers_lb).handover_connection(con, data)
-            _LOGGER.debug("Handover new peer connection: %s", data)
-            return None
+        if client.buffer.startswith(b"gA"):
+            next(workers_lb).handover_connection(client.sock, client.buffer)
+            _LOGGER.debug("Handover new peer connection: %s", client.buffer)
+            client.soft_close()
```
Comment (ludeeus) [inline snitun/server/run.py:331]: Shouldn't this happen before the handover?
Like before?

File: snitun/server/run.py [hunk 14]
```diff
@@ -162,6 +154,35 @@ async def _handler(
             return
 
 
+@attr.s(slots=True)
+class Connection:
+    """Partial data class."""
```
Comment (ludeeus) [inline snitun/server/run.py:159]:
```suggestion
    """Connection data class."""
```

File: snitun/server/run.py [hunk 15]
```diff
@@ -260,39 +269,35 @@ def _register_parial(fileno: int, data: bytes) -> bool:
...
             # cleanup stale connection
-            for fileno in tuple(stale):
-                if fileno not in connections:
-                    stale.pop(fileno)
-                elif stale[fileno] >= WORKER_STALE_MAX:
-                    self._poller.unregister(fileno)
-                    con = connections.pop(fileno)
-                    partial.pop(fileno, None)
-                    self._close_socket(con)
+            for client_id in list(connections.keys()):
```
Comment (ludeeus) [inline snitun/server/run.py:294]: Tuple is faster, why change that?

File: snitun/server/run.py (hunk 12 above)
Comment (pvizeli) [reply snitun/server/run.py:359]: Doesn't matters as we take it out of the epoll before the next circle. But I can change it

File: snitun/server/run.py (hunk 13 above)
Comment (pvizeli) [reply snitun/server/run.py:331]: dito

**Pull request closed without merging**